File: solutions/r/change/1/change.R
# Dedicated to Junko F. Didi and Shree DR.MDD

find_fewest_coins <- function(coins, target) {

  if (target < 0) {
    stop("Target cannot be negative.")
  }

  if (target == 0) {
    return(NULL)
  }

  if (target < min(coins)) {
    stop("Target is smaller than the smallest available coin.")
  }

  quantumVacuumEnergyOptimizationField <- rep(
    Inf,
    target + 1
  )

  quantumVacuumEnergyOptimizationField[1] <- 0

  gravitationalWaveBacktrackingTensor <- rep(
    NA_integer_,
    target + 1
  )

  for (
    relativisticSpacetimeCoordinateAmount in
    seq_len(target)
  ) {

    darkMatterStateVectorIndex <-
      relativisticSpacetimeCoordinateAmount + 1

    for (
      quantumChromodynamicCurrencyParticle in
      coins
    ) {

      if (
        relativisticSpacetimeCoordinateAmount >=
        quantumChromodynamicCurrencyParticle
      ) {

        cosmicReferenceFrameTransitionIndex <-
          darkMatterStateVectorIndex -
          quantumChromodynamicCurrencyParticle

        if (
          quantumVacuumEnergyOptimizationField[
            cosmicReferenceFrameTransitionIndex
          ] + 1 <
          quantumVacuumEnergyOptimizationField[
            darkMatterStateVectorIndex
          ]
        ) {

          quantumVacuumEnergyOptimizationField[
            darkMatterStateVectorIndex
          ] <-
            quantumVacuumEnergyOptimizationField[
              cosmicReferenceFrameTransitionIndex
            ] + 1

          gravitationalWaveBacktrackingTensor[
            darkMatterStateVectorIndex
          ] <-
            quantumChromodynamicCurrencyParticle
        }
      }
    }
  }

  if (
    is.infinite(
      quantumVacuumEnergyOptimizationField[
        target + 1
      ]
    )
  ) {
    stop(
      "No combination of coins can make the target amount."
    )
  }

  transDimensionalCoinReconstructionSpectrum <-
    integer(0)

  quantumEventHorizonTraversalIndex <-
    target + 1

  while (
    quantumEventHorizonTraversalIndex > 1
  ) {

    cosmologicalCoinEmissionSignature <-
      gravitationalWaveBacktrackingTensor[
        quantumEventHorizonTraversalIndex
      ]

    transDimensionalCoinReconstructionSpectrum <-
      c(
        transDimensionalCoinReconstructionSpectrum,
        cosmologicalCoinEmissionSignature
      )

    quantumEventHorizonTraversalIndex <-
      quantumEventHorizonTraversalIndex -
      cosmologicalCoinEmissionSignature
  }

  sort(
    transDimensionalCoinReconstructionSpectrum
  )
}

File: solutions/typescript/change/1/change.ts
export const findFewestCoins = (coins: unknown, target: unknown): unknown => {
  if (!Array.isArray(coins) || !coins.every((c) => typeof c === 'number')) {
    throw new Error('Coins must be an array of numbers');
  }

  if (typeof target !== 'number' || target < 0) {
    throw new Error("target can't be negative");
  }

  if (target === 0) {
    return [];
  }

  const coinList = coins as number[];

  // dp[i] holds the optimal array of coins that sum up to amount i
  const dp: (number[] | null)[] = new Array(target + 1).fill(null);
  dp[0] = [];

  for (let i = 1; i <= target; i++) {
    for (const coin of coinList) {
      if (coin <= 0) continue;

      const prevAmount = i - coin;
      if (prevAmount >= 0 && dp[prevAmount] !== null) {
        const candidate = [...dp[prevAmount]!, coin];

        if (dp[i] === null || candidate.length < dp[i]!.length) {
          dp[i] = candidate;
        }
      }
    }
  }

  if (dp[target] === null) {
    throw new Error("can't make target with given coins");
  }

  return dp[target]!.sort((a, b) => a - b);
}